src/box/box.cc

```diff
@@ -3201,6 +3202,25 @@ box_session_push(const char *data, const char *data_end)
 	return rc;
 }
 
+API_EXPORT int
+box_iproto_send(uint64_t sid,
+		const char *header, const char *header_end,
+		const char *body, const char *body_end)
+{
+	struct session *session = session_find(sid);
+	if (session == NULL) {
+		diag_set(ClientError, ER_NO_SUCH_SESSION, sid);
+		return -1;
+	}
+	if (session->type != SESSION_TYPE_BINARY) {
+		const char *name = tt_sprintf("Session '%s'",
+					      session_type_strs[session->type]);
+		diag_set(ClientError, ER_UNSUPPORTED, name, "send()");
+		return -1;
+	}
+	return iproto_session_send(session, header, header_end, body, body_end);
```

Comment:
I think we can specify this is as UB.

@locker @Gerold103 RFC

Reply:
Better not leave UB here. Especially in user-visible API.
AFAIU every time someone finds a way to lead tarantool to a segfault (here it would be dereferencing a NULL header, for example), this is considered a bug on our side and fixed. Let's not introduce such bugs where possible.

Reply:
But actually our implementation can already invoke UB by calculating the pointer difference (when the two pointer are not pointer to contiguous memory).

Reply:
AFAIK there are no such checks in tarantool API at all. See [box_select](https://github.com/tarantool/tarantool/blob/master/src/box/box.cc#L2846) for example. I think there's no sense in such a check only in `iproto_session_send`.

Reply:
It is now possible to cause sigsigv via C API. That's ok.
But it should not be possible to cause sigsegv using Lua API.

Reply:
@alyapunov
> AFAIK there are no such checks in tarantool API at all. See [box_select](https://github.com/tarantool/tarantool/blob/master/src/box/box.cc#L2822) for example. I think there's no sense in such a check only in iproto_session_send.

I don't see any problem with box_select at first glance. For example, the case when `key` is NULL is handled just fine [here](https://github.com/tarantool/tarantool/blob/2ed06651c317f2d46631b0d55e4b88f2c8a6c0a0/src/box/box.cc#L2846)

Anyway, if you people are saying this is fine, I won't object.

Reply:
As I said, the case when the parameters are NULL is perfectly handled here.

Reply:
What do you mean by "one parameter is NULL and the other isn't."?

If the both `parameter` and `parameter_end` are `NULL`, everything is handled correctly (otherwise, how do the tests with empty packet bodies pass)?

If `parameter` is `NULL` and `parameter_end` is not NULL or vice versa, then the user is obviously shooting himself in the foot.